### ISSUE TITLE
Race condition with save/delete (tested on 0.32.2)?

### ISSUE THREAD

**Issue by jmmills** [josiahcarlson/rom#73]
Here is a situation in which rom will save an incomplete object in redis and then later throw an exception because required columns are missing in the data store.

* Process A creates a object, let's say it has id 9
* Process B fetches data for id 9 and inflates it into an object
* Process B modifies a property of the object
* Process A calls delete() on object (id:9)
* Process B calls save() on object (id:9)
* Another get() is called for id 9 at a later time throws an exception since the incomplete object was saved back to redis

It seems like rom should not allow you to save() an object that has already had delete() called elsewhere. Or maybe it's not rom's problem to be thread safe, and it's up to me to make sure I don't run into these race conditions... not sure, but I thought I would at least throw out this sample/test code and get some input.

Thoughts?

```
import time
import rom
import threading

from nose import *

class Thing(rom.Model):
    a = rom.String(required=True, index=True, keygen=rom.SIMPLE)
    b = rom.String(default='foo', index=True, keygen=rom.SIMPLE)


def get_obj_wait_then_save(ev, i):
    o = Thing.get(i)

    while not ev.isSet():
        pass

    o.b = 'bar'
    o.save()


def test_race_condition():
    o = Thing(a='import value')
    assert o.save()

    f_id = o.id

    e = threading.Event()
    t = threading.Thread(name='non-block',
                         target=get_obj_wait_then_save,
                         args=(e, o.id))

    t.start()
    time.sleep(.01)

    o.delete()
    assert not Thing.get(f_id)

    e.set()
    time.sleep(.01)

    assert not Thing.get(f_id)
```
The failed test output:
```
E
======================================================================
ERROR: rom_thread_race.test_race_condition
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/jmills/virtualenv/lib/python2.7/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/Users/jmills/rom_thread_race.py", line 42, in test_race_condition
    assert not Thing.get(f_id)
  File "/Users/jmills/virtualenv/lib/python2.7/site-packages/rom/model.py", line 565, in get
    out[i] = cls(_loading=True, **data)
  File "/Users/jmills/virtualenv/lib/python2.7/site-packages/rom/model.py", line 230, in __init__
    setattr(self, attr, data)
  File "/Users/jmills/virtualenv/lib/python2.7/site-packages/rom/columns.py", line 287, in __set__
    self._init_(obj, *value)
  File "/Users/jmills/virtualenv/lib/python2.7/site-packages/rom/columns.py", line 270, in _init_
    raise MissingColumn("%s.%s cannot be missing"%(self._model, self._attr))
MissingColumn: Thing.a cannot be missing

----------------------------------------------------------------------
Ran 1 test in 0.029s

FAILED (errors=1)
```

**Comment by josiahcarlson** [comment]
Long story short, this is a data versioning problem. I have been aware of this and several other related issues for a while, but hadn't spent the 5 minutes convincing myself that it was okay for `.save()` and `session.commit()` to raise an exceptions. I am now convinced and will do some work to fix this tonight.

**Comment by josiahcarlson** [comment]
I have a fix for this and will release an updated version of rom in the next few days.

**Comment by jmmills** [comment]
Awesome, glad to be of help. What would the logic be to check to see if an item has already been deleted?

**Comment by josiahcarlson** [comment]
To handle the issue of *any* column being modified by two different clients at the same time, I'm adding a check to verify that any old data in the entity is the same prior to writing. If any prior data has changed, you get a new `DataRaceError()` exception.

To check the deleted case, we do the exact same thing, only also add a check to ensure that the primary key is stored in the entity data. In the case that the entity does not exist, we raise a `EntityDeletedError()` exception, requiring that the user use `entity.save(force=True)` to re-save the entire entity.

This is now available in version 0.33.0, available now in Github and PyPI.
